File: LRU_Cache/lru_cache.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.next = None
        self.prev = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}

        # head = LRU, tail = MRU
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head
        

    def get(self, key: int) -> int:
        # if self.cache is empty, return -1
        if self.cache == {}:
            return -1

        if key in self.cache:
            # REMEMBER TO UPDATE POINTERS
            # remove it and insert at end of DLL
            node = self.cache[key]
            self.remove(node)
            self.insert(node)
            return node.value
  
        return -1


        #imagine we're changing the Node that is in the middle of a list.
        # Don't worry about head or that stuff. That's taken care of in the constructor
    def remove(self, node):
        prv = node.prev #identifying the Nodes' whose pointers we want to update
        nxt = node.next 
        prv.next = nxt
        nxt.prev = prv

    def insert(self, node):
        prv = self.tail.prev # identify the Nodes whose pointers we want to update
        nxt = self.tail
        
        prv.next = node
        nxt.prev = node
        
        node.prev = prv               
        node.next = nxt


    def put(self, key: int, value: int) -> None:
        
        if key in self.cache:
            self.remove(self.cache[key])
        
        new_node = Node(key, value)
        self.cache[key] = new_node
        self.insert(new_node)


        if len(self.cache) > self.capacity:
            lru = self.head.next
            self.remove(lru)
            del self.cache[lru.key]
            

 


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: LRU_Cache/lru_cache.cpp
#include <unordered_map>
#include <iostream>
using namespace std;

class Node {
    public:
        int key;
        int value;
        Node* next;
        Node* prev;
    
        Node(int k, int v) : key(k), value(v), next(nullptr), prev(nullptr) {}
    
    
    };
    
    class LRUCache {
    
    private:
        int capacity;
        std::unordered_map<int, Node*> cache;
    
        Node* head;
        Node* tail;
    
        void remove(Node* node) {
            Node* prev = node->prev;
            Node* next = node->next;
            prev->next = next;
            next->prev = prev;
        }
    
        void insert(Node* node) {
            Node* prev = tail->prev;
            Node* next = tail;
    
            prev->next = node;
            next->prev = node;
    
            node->prev = prev;
            node->next = next;
        }
    
    public:
        LRUCache(int capacity) : capacity(capacity) {
            head = new Node(0, 0);
            tail = new Node(0, 0);
            head->next = tail;
            tail->prev = head;
        }
        ~LRUCache() {
            Node* current = head;
            while (current != nullptr) {
                Node* next = current->next;
                delete current;
                current = next;
            }
        }
        
        int get(int key) {
            if (cache.empty()) {
                return -1;
            }
    
            auto it = cache.find(key);
            if (it != cache.end()) {
                Node* node = it->second;
                remove(node);
                insert(node);
                return node->value;
            }
            return -1;
            
        }
        
        void put(int key, int value) {
            auto iter = cache.find(key);
            if (iter != cache.end()) {
                remove(iter->second);
            }
    
            Node* newNode = new Node(key, value);
            cache[key] = newNode;
            insert(newNode);
    
            if (cache.size() > capacity) {
                Node* lru = head->next;
                remove(lru);
                cache.erase(lru->key);
                delete lru;
            }
        }
    };
    
    /**
     * Your LRUCache object will be instantiated and called as such:
     * LRUCache* obj = new LRUCache(capacity);
     * int param_1 = obj->get(key);
     * obj->put(key,value);
     */